javascript/javascript2/week2/Week2HW.js

```diff
@@ -0,0 +1,126 @@
+console.log(averageRatedMovies);
+const badRatedMovies = movies
+  .filter((movie) => movie.rating < 4)
+  .map((movie) => Object.assign(movie, { tag: "Bad" }));
+console.log(badRatedMovies);
+console.log(movies);
+
+// Filtering movies who have ratings higher than 6 and creating the array with only one key-value of rating
+const moviesHighRated = movies
+  .filter((movie) => movie.rating > 6)
+  .map((movie) => movie.rating);
+console.log(moviesHighRated);
+
+// counting the total number of movies which contains the keywords (surfer ,alien or Benjamin) and  it shouldn't be case sensitive
+
+const moviesWithKeywords = movies.filter(
+  (movie) =>
+    movie.title.toLocaleLowerCase().includes("surfer".toLocaleLowerCase()) ||
+    movie.title.toLocaleLowerCase().includes("alien".toLocaleLowerCase()) ||
+    movie.title.toLocaleLowerCase().includes("benjamin".toLocaleLowerCase())
+);
+console.log(
+  `The number of the movies which contain any of the keywords (Surfer, Alien and Benjamin) is ${moviesWithKeywords.length}`
+);
+
+//finding duplicated words
+
+// let duplicateWords = new RegExp(/\b(\w+)\b.*\b\1\b/i);
+// let duplicateWordMovies = movies
+//   .filter((movie) => movie.title.match(duplicateWords))
+//   .map((movie) => movie.title);
+// console.log(duplicateWordMovies.length);
+
+const duplicatedWord = movies.filter((currentMovie) => {
+  const currentTitle = currentMovie.title.toLowerCase().split(" ");
+  const sameWords = currentTitle.filter(
+    (word, index) => currentTitle.indexOf(word) !== index
+  );
+  if (sameWords.length > 0) {
+    return currentMovie;
+  }
+});
+console.log(`The number of the duplicated words is ${duplicatedWord.length}`);
+
+// finding the average rating
+
+const moviesRateOnly = movies
+  .filter((movie) => movie.rating)
+  .map((movie) => movie.rating);
+// console.log(moviesRateOnly);
+const averageRate = moviesRateOnly.reduce(function (a, b) {
+  const totalRate = a + b / moviesRateOnly.length;
+  return totalRate;
+}, 0);
+console.log(`The average of the rating of all the movies is ${averageRate}`);
+
+//Count the total number of Good, Average and Bad movies
+
+//counting total of good movies
+const goodMovies = goodRatedMovies.map((movie) => movie.rating >= 7);
+const totalGoodMovies = goodMovies.reduce(function (c, d) {
+  return c + d;
+}, 0);
+console.log(`Total number of Good movies are ${totalGoodMovies}`);
+
+// counting total of average movies
+const averageMovies = averageRatedMovies.map(
+  (movie) => movie.rating < 7 && movie.rating >= 4
+);
+const totalAverageMovies = averageMovies.reduce(function (e, f) {
+  return e + f;
+}, 0);
+console.log(`Total number of Average movies are ${totalAverageMovies}`);
+
+// counting total of Bad movies
+const badMovies = badRatedMovies.map((movie) => movie.rating < 4);
+const totalBadMovies = badMovies.reduce(function (g, h) {
+  return g + h;
+}, 0);
+console.log(`Total number of Bad movies are ${totalBadMovies}`);
```

Comment:
No need of maping here as  badRatedMovies, averageRatedMovies, goodRatedMovies already have `tag  ` key.
I think you have created "tag" key (value: Good, Average and Bad ) in `movies `array already.  It is more efficient to do this way. (Note: In your case `moviesWithTag `is `movies`)
```
const ratingFreq1 = moviesWithTag.reduce( (freq, movie) => {
    freq[movie.tag]++;
    return freq;   
},{'Good': 0, 'Average': 0, 'Bad': 0});
```